src/zkevm_specs/pi_circuit.py

```diff
@@ -91,62 +137,43 @@ class PublicInputs:
 def check_row(
     row: Row,
     row_next: Row,
-    row_offset_block_table_value_hi: Row,
-    row_offset_tx_table_tx_id: Row,
-    row_offset_tx_table_index: Row,
-    row_offset_tx_table_value_lo: Row,
-    row_offset_tx_table_value_hi: Row,
-    table: Set[TxCallDataGasCostAccRow],
+    calldata_gas_cost_table: Set[TxCallDataGasCostAccRow],
     fixed_u16_table: Set[FixedU16Row],
+    keccak_table: KeccakTable,
+    circuit_len: FQ,
 ):
-    q_not_end = row.q_not_end
-    q_end = row.q_end
-    row_offset_block_table_value_lo = row
-
-    # 0.0 rpi_rlc_acc[0] == RLC(raw_public_inputs, rand_rpi)
-    assert q_not_end * row.rpi_rlc_acc == q_not_end * (
-        row_next.rpi_rlc_acc * row.rand_rpi + row.raw_public_inputs
-    )
-
-    assert q_end * row.rpi_rlc_acc == q_end * row.raw_public_inputs
+    q_bytes_last = row.q_bytes_last
+    q_rpi_byte_enable = row.q_rpi_byte_enable
 
-    # 0.1 rand_rpi[i] == rand_rpi[j]
-    assert q_not_end * row.rand_rpi == q_not_end * row_next.rand_rpi
+    # gate 1 and gate 2 are compensation branch
+    # 1: rpi_bytes_keccakrlc[last] = rpi_bytes[last]
+    assert q_rpi_byte_enable * q_bytes_last * (row.rpi_bytes_keccakrlc - row.rpi_bytes) == FQ(0)
 
-    # 0.2 Block table -> value column match with raw_public_inputs at expected offset
-    assert (
-        row.q_block_table * row.block_table.value.lo.expr()
-        == row.q_block_table * row_offset_block_table_value_lo.raw_public_inputs
-    )
+    # 2: rpi_bytes_keccakrlc[i] = keccak_rand * rpi_bytes_keccakrlc[i+1] + rpi_bytes[i]"
     assert (
-        row.q_block_table * row.block_table.value.hi.expr()
-        == row.q_block_table * row_offset_block_table_value_hi.raw_public_inputs
+        q_rpi_byte_enable
+        * (FQ(1) - q_bytes_last)
+        * (row.rpi_bytes_keccakrlc - row_next.rpi_bytes_keccakrlc * keccak_rand - row.rpi_bytes)
     )
 
-    # 0.3 Tx table -> {tx_id, index, value} column match with raw_public_inputs at expected offset
-    assert (
-        row.q_tx_table * row.tx_table.tx_id
-        == row.q_tx_table * row_offset_tx_table_tx_id.raw_public_inputs
-    )
-    assert (
-        row.q_tx_table * row.tx_table.index
-        == row.q_tx_table * row_offset_tx_table_index.raw_public_inputs
-    )
-    assert (
-        row.q_tx_table * row.tx_table.value.lo.expr()
-        == row.q_tx_table * row_offset_tx_table_value_lo.raw_public_inputs
-    )
-    assert (
-        row.q_tx_table * row.tx_table.value.hi.expr()
-        == row.q_tx_table * row_offset_tx_table_value_hi.raw_public_inputs
-    )
-    assert (
-        row.q_tx_calldata * row.tx_table.value.lo.expr()
-        == row.q_tx_calldata * row_offset_tx_table_value_lo.raw_public_inputs
+    # gate 3 and gate 4 are compensation branch
+    # 3: rpi_value_lc[i] = rpi_value_lc[i+1] * byte_pow_base + rpi_bytes[i]
```

Comment:
Hi
Here the `q_rpi_value_start` is used to mark the split between values.

Layout example as below (Thanks for remind, I will add those in comment to make idea easily to understand)
```
| q_rpi_value_start | rpi_byte | rpi_value_lc       | memo       |
| ----------------- | -------- | ------------------ |------------|
| ...               | .....    | ...     ...        |            |
|                   | dd       | ddee               |gate 3      |
| 1                 | dd       | dd                 |gate 4      | 
|                   | cc       | aabbcc             |gate 3      |
|                   | bb       | aabb               |gate 3      |
| 1                 | aa       | aa                 |gate 4      |
```

Reply:
In `specs/public_inputs.md` we plan to adapt https://eips.ethereum.org/EIPS/eip-4844 so verifier even can save the effort to call precompile keccak. Keccak is more like a transient plan so that's why I didn't touch md specs this time and plan to revamp it later. But it make sense to refresh some out-of-date part to reflect the idea now. I just update few description in MD specs